go/interpreter/lfvm/instructions_test.go

```diff
@@ -2036,6 +1912,86 @@ func TestOpExtCodeHash_WritesHashOnStackIfAccountExists(t *testing.T) {
 	}
 }
 
+func TestInstructions_MCopy_ReturnsErrorOnFailure(t *testing.T) {
+
+	tests := map[string]struct {
+		destOffset, offset, size uint64
+		expectedError            error
+		gasRemoved               uint64
+	}{
+		"does nothing if size is zero": {
+			destOffset: 1024, offset: 125, size: 0,
+		},
+		"returns error when failed read memory expansion": {
+			destOffset: 0, offset: math.MaxUint64, size: 1,
+			expectedError: errOverflow,
+		},
+		"returns error when failed write memory expansion": {
+			destOffset: math.MaxUint64, offset: 0, size: 1,
+			expectedError: errOverflow,
+		},
+		"returns error if gas is not enough for size": {
+			destOffset: 0, offset: 0, size: 128,
+			expectedError: errOutOfGas,
+			gasRemoved:    1,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			ctxt.params.Revision = tosca.R13_Cancun
+			ctxt.stack = fillStack(
+				*uint256.NewInt(test.destOffset),
+				*uint256.NewInt(test.offset),
+				*uint256.NewInt(test.size),
+			)
+
+			// ignore memory setup errors, to focus on the mcopy operation
+			// expansion is done to accumulate memory cost and focus on the
+			// word count gas cost.
+			_ = ctxt.memory.expandMemory(test.destOffset, test.size, &context{gas: 1 << 32})
+			_ = ctxt.memory.expandMemory(test.offset, test.size, &context{gas: 1 << 32})
+			ctxt.gas = tosca.Gas(3*tosca.SizeInWords(test.size) - test.gasRemoved)
```

Comment:
this is unnecessary. if you set `ctxt.gas` to 6, (for the case of write memory overflow), and the size of the "out of gas" case to 64, the tests still works as intended, but with less code and you can remove `gasRemoved` from tests struct.

Reply:
Same issue as the memory, 
The test code is generic and works with any input test case, if we hardcode gas here two things may happen:
we cannot execute new test cases
we and up triggering out of gas for a different reason than the one the test intends to address. 

This test is modelled to exploit precisely the scenario where gas is less than the size, provided by the parameters of the test.